ql/src/java/org/apache/hadoop/hive/ql/hooks/OperatorHealthCheckerHook.java

```diff
@@ -0,0 +1,211 @@
+
+import com.google.common.collect.Sets;
+import org.apache.hadoop.hive.ql.exec.ColumnInfo;
+import org.apache.hadoop.hive.ql.exec.Operator;
+import org.apache.hadoop.hive.ql.exec.ReduceSinkOperator;
+import org.apache.hadoop.hive.ql.exec.RowSchema;
+import org.apache.hadoop.hive.ql.exec.ScriptOperator;
+import org.apache.hadoop.hive.ql.exec.SelectOperator;
+import org.apache.hadoop.hive.ql.exec.TableScanOperator;
+import org.apache.hadoop.hive.ql.exec.Task;
+import org.apache.hadoop.hive.ql.lib.DefaultGraphWalker;
+import org.apache.hadoop.hive.ql.lib.DefaultRuleDispatcher;
+import org.apache.hadoop.hive.ql.lib.SemanticDispatcher;
+import org.apache.hadoop.hive.ql.lib.SemanticGraphWalker;
+import org.apache.hadoop.hive.ql.lib.Node;
+import org.apache.hadoop.hive.ql.lib.SemanticNodeProcessor;
+import org.apache.hadoop.hive.ql.lib.NodeProcessorCtx;
+import org.apache.hadoop.hive.ql.parse.SemanticException;
+import org.apache.hadoop.hive.ql.plan.BaseWork;
+import org.apache.hadoop.hive.ql.plan.ExprNodeConstantDesc;
+import org.apache.hadoop.hive.ql.plan.ExprNodeDesc;
+import org.apache.hadoop.hive.ql.plan.GroupByDesc;
+import org.apache.hadoop.hive.ql.plan.MapWork;
+import org.apache.hadoop.hive.ql.plan.MapredWork;
+import org.apache.hadoop.hive.ql.plan.OperatorDesc;
+import org.apache.hadoop.hive.ql.plan.ReduceWork;
+import org.apache.hadoop.hive.ql.plan.SelectDesc;
+import org.apache.hadoop.hive.ql.plan.TezWork;
+
+/**
+ * Checks some operator quality rules.
+ *
+ * Checks whenever operator ids are not reused.
+ * Checks some level of expression/schema consistency
+ * Some sanity checks on SelectOperators
+ */
+public class OperatorHealthCheckerHook implements ExecuteWithHookContext {
+
+  static class UniqueOpIdChecker implements SemanticNodeProcessor {
+
+    Map<String, Operator<?>> opMap = new HashMap<>();
+
+    @Override
+    public Object process(Node nd, Stack<Node> stack, NodeProcessorCtx procCtx, Object... nodeOutputs)
+        throws SemanticException {
+
+      Operator<?> op = (Operator<?>) nd;
+      checkOperator(op);
+      String opKey = op.getOperatorId();
+      Operator<?> found = opMap.get(opKey);
+      if (found != null) {
+        throw new RuntimeException("operator id reuse found: " + opKey);
+      }
+      opMap.put(opKey, op);
+      return null;
+    }
+  }
+
+  public static void checkOperator(Operator<?> op) {
+    OperatorDesc conf = op.getConf();
+    Map<String, ExprNodeDesc> exprMap = conf.getColumnExprMap();
+    RowSchema schema = op.getSchema();
+
+    chceckSchema(schema);
+    if (op instanceof SelectOperator) {
+      checkSelectOperator((SelectOperator) op);
+    }
+    if (schema != null && exprMap != null) {
+      for (Entry<String, ExprNodeDesc> c : exprMap.entrySet()) {
+        if (c.getValue() instanceof ExprNodeConstantDesc) {
+          continue;
+        }
+        ColumnInfo ci = schema.getColumnInfo(c.getKey());
+        if (c.getKey().startsWith("KEY.reducesinkkey")) {
```

Comment:
I don't think KEY identifier is ever going to change, but you could use `Utilities.ReduceField.KEY + ".reducesinkkey"`.

Reply:
actually - this if is still here because in case of MJ the RS rowschema is so bad that I kinda give up fixing it ; it needs to be taken care of separately...

Reply:
opened HIVE-24799